contracts/extensions/VotingReputation.sol

```diff
@@ -0,0 +1,768 @@
+/*
+  This file is part of The Colony Network.
+
+  The Colony Network is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  The Colony Network is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with The Colony Network. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+pragma solidity 0.5.8;
+pragma experimental ABIEncoderV2;
+
+import "./../../lib/dappsys/math.sol";
+import "./../colony/ColonyDataTypes.sol";
+import "./../colony/IColony.sol";
+import "./../colonyNetwork/IColonyNetwork.sol";
+import "./../common/ERC20Extended.sol";
+import "./../patriciaTree/PatriciaTreeProofs.sol";
+import "./../tokenLocking/ITokenLocking.sol";
+
+
+contract VotingReputation is DSMath, PatriciaTreeProofs {
+
+  // Events
+  event ExtensionInitialised();
+  event ExtensionDeprecated();
+  event PollCreated(uint256 indexed pollId, uint256 indexed skillId);
+  event PollStaked(uint256 indexed pollId, address indexed staker, bool indexed side, uint256 amount);
+  event PollVoteSubmitted(uint256 indexed pollId, address indexed voter);
+  event PollVoteRevealed(uint256 indexed pollId, address indexed voter, bool indexed side);
+  event PollExecuted(uint256 indexed pollId, bytes action, bool success);
+  event PollRewardClaimed(uint256 indexed pollId, address indexed staker, bool indexed side, uint256 amount);
+
+  // Constants
+  uint256 constant UINT256_MAX = 2**256 - 1;
+  uint256 constant NAY = 0;
+  uint256 constant YAY = 1;
+
+  uint256 constant STAKE_PERIOD = 3 days;
+  uint256 constant VOTE_PERIOD = 2 days;
+  uint256 constant REVEAL_PERIOD = 2 days;
+
+  bytes4 constant CHANGE_FUNCTION = bytes4(
+    keccak256("setExpenditureState(uint256,uint256,uint256,uint256,bool[],bytes32[],bytes32)")
```

Comment:
This [figma](https://www.figma.com/file/JIlC5rbmv01G5pQMMtpAPboJ/Voting?node-id=1095%3A46455) shows the intention for everything that should be able to be done with this extension, so I think it's important that we get the scaffolding as general as possible at this point. 

In particular, I think we need to work out a way of handling 'duplicate actions' in a much more generalised manner - I don't think this approach is going to scale terribly well or have the behaviour that we want. This might require us to hold our colony interface to a higher standard ('always ignore the last slot when determining duplicates), or perhaps adding metadata to the resolver of the contracts involved (a mapping that was `function sig -> byte array to & with action before hashing', for example, might get us a long way).

At the bare minimum, for the first version, it needs to be able to handle moving funds between pots, and oneTxPayments with 'appropriate' duplicate protection, whatever we decide that means; this is perhaps a good thing, forcing us to consider both colony interactions and extension interactions; perhaps some tests with oneTxPayments will help us get our house in order on that front.

Reply:
It's a great point, the problem of a general way of managing concurrent actions doesn't seem to have an obviously great solution. I think that pulling out the expenditure state change function as a special case is appropriate, since it's the only one which leads to payouts and is (likely) the main source of dispute. For other functions, I suppose carving out special behavior on a case-by-case basis might be the best way.

Something to think about -- should some functions *not* consider identical actions be duplicates? I.e. if I want to move 10 tokens from pot A to pot B, and then someone else also wants to move 10 tokens from pot A to pot B, who are we to say that that's a "duplicate"?

Your idea about storing some function-specific metadata is interesting... it means that changes to colony functions will get automatically picked up by the voting extension without needing to upgrade the latter. Maybe something to have a call about?

Reply:
So I had a call with Jack to try and think about this... I think the rule of thumb that we settled on was that if an action is doing or creating something new, then duplicates should be allowed. So in your example of moving tokens between pots, that should not be considered a duplicate, and should be allowed. If an action is altering something that already exists, then duplicates should not be allowed - so, for example, the amount being paid out by an expenditure.

Where it gets a bit more nebulous is multiple votes on an expenditure. If there are two votes on an expenditure, one about who it's being paid out to, and one about how much is being paid out... you could make an argument that those shouldn't be able to happen at the same time. But if they were two different slots in the expenditure (one for a worker, one for how much the evaluator is being paid) then there's no problem. And moreover, that involves the voting contract knowing about all the different things built on top of expenditures. I think that is a huge nest of rats to try and handle though, so we may just have to make do...

In conclusion: I definitely think a call to try and hash some of this out is going to be best :smile:

Reply:
Some thoughts:

- Only need to worry about permissioned functions anyway
- See what the special cases are (root, non-root, duplicate protection or not)
- Might be possible to figure this out without an exhaustive list

Reply:
Another angle here, worth mentioning, is that any "locking" function creates an opportunity for a griefing attack where I can create pointless motions to inhibit a valid motion. In a sense the existing incentives make this a non-issue since I'll likely lose my stake by losing the vote, but there may be scenarios where that cost is worth it. Alex mentioned this w.r.t. taking away permissions, so I wanted to make a note here.

Reply:
Thinking about it more, I'm not sure if we're doing ourselves any favors by thinking about motion locking apart from the special case of expenditure state change
And even then, it's more about locking the expenditure from being claimed then about preventing competing motions
Frivolous motions are already disincentivized via stakes
So having a second mechanism for preventing motion spamming seems unecessary
Can't think of anything more general than "no locking" after all :slight_smile:
The reason we lock expenditures isn't to prevent competing motions anyway, it's to prevent funds being claimed while a motion is ongoing
Which is very different from preventing simultaneous competing motions